File: leetcode/copyListWithRandomPointer/tom/copyListWithRandomPointer.py
"""
A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

Return a deep copy of the list.

 

Example 1:



Input:
{"$id":"1","next":{"$id":"2","next":null,"random":{"$ref":"2"},"val":2},"random":{"$ref":"2"},"val":1}

Explanation:
Node 1's value is 1, both of its next and random pointer points to Node 2.
Node 2's value is 2, its next pointer points to null and its random pointer points to itself.
 

Note:

You must return the copy of the given head as a reference to the cloned list


"""

"""

basic idea : we just go through the linked list and create every node,
and store the corresponding node to the hashTable.

Then, we just go through the list and  connnect the random pointer to the corresponding node 


"""



"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random
"""
class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """

        memo = {}
        first = 1
        tmp = head
        newHead = None
        newTmp = None
        while tmp :                ### just go throght the list and create every node 
            if first == 1 :
                first += 1
                newHead = Node(tmp.val,None,None)
                newTmp = newHead
                memo[tmp] = newTmp
                tmp = tmp.next
                continue
            else :
                newTmp.next = Node(tmp.val,None,None)
                newTmp = newTmp.next
                memo[tmp] = newTmp
                tmp = tmp.next

        tmp = head
        newTmp = newHead
        while tmp : ### go through the list again, and connect the random pointer
            if tmp.random :
                newTmp.random = memo[tmp.random]
            else :
                newTmp.random = None
            newTmp = newTmp.next
            tmp = tmp.next
        return (newHead)

File: leetcode/copyListWithRandomPointer/guanduo/copyListWithRandomPointer.cpp
/*
A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

Return a deep copy of the list.

*/


/*

// Definition for a Node.
class Node {
public:
  int val;
  Node* next;
  Node* random;

  Node() {}

  Node(int _val, Node* _next, Node* _random) {
    val = _val;
    next = _next;
    random = _random;
  }
};
*/


//Use recursion to solve the problem
//If there is next, keep copying nodes till end. Then if there is random, keep copying nodes till end

class Solution {
private:
  unordered_map<Node*,Node*> mappingT;              //make it a private member to keep code clean.key is original list, item is the new list node
public:
  Node* copyRandomList(Node* head) {
    if(head == NULL) return NULL;                   //empty list or we have reached end
    if(mappingT.count(head)) return mappingT[head]; //if current node has been constructed before (either through random or next, no need to copy again
    Node* newNode = new Node(head->val);            //got a new node. construct it 
    mappingT[head] = newNode;                       //record the node to the map.
    newNode->next = copyRandomList(head->next);     //recursively copy next nodes
    newNode->random = copyRandomList(head->random); //recursively copy random nodes
    return newNode;
  }

};
